ScrollViewer expression animations extensions

Sergio0694 opened the pull request:
This PR is part of the group being tracked in #3108.

<!-- Add a brief overview here of the feature/bug & fix -->

## PR Type
What kind of change does this PR introduce?
<!-- Please uncomment one ore more that apply to this PR -->

<!-- - Bugfix -->
 - Feature
<!-- - Code style update (formatting) -->
<!-- - Refactoring (no functional changes, no api changes) -->
<!-- - Build or CI related changes -->
<!-- - Documentation content changes -->
<!-- - Sample app changes -->
<!-- - Other... Please describe: -->

## What is the new behavior?
<!-- Describe how was this issue resolved or changed? -->
Two extension methods for the `ScrollViewer` class that can be used to easily create and start expression animations that bind the scrolling translation on a specified axis of the source scroller, to the offset on a given axis of a target control. This can be used to keep "in sync" the contents of a `ScrollViewer` with some overlays rendered on another surface that is not part of the `ScrollViewer`, as the expression animation will adjust the offset of the external surface to perfectly follow its content.

Here it is in action, you can see that there is some text inside a `RichEditBox`, and these new extensions are used to keep the text in sync with the text overlays (the dashed column guides and the TAB arrows) and the indentation indicators on the left, which are all rendered in a Win2D canvas outside of the `ScrollViewer` within the text control. Additionally, the line number are displayed through a `TextBlock` living in a `Canvas`, which is also kept in sync with the text in the same way:

![KyjiuIca5M](https://user-images.githubusercontent.com/10199417/78357926-9c972100-75b2-11ea-81c2-9fa0348a794b.gif)


## PR Checklist

Please check if your PR fulfills the following requirements:

- [X] Tested code with current [supported SDKs](../readme.md#supported)
- [ ] Pull Request has been submitted to the documentation repository [instructions](..\contributing.md#docs). Link: <!-- docs PR link -->
- [X] Sample in sample app has been added / updated (for bug fixes / features)
    - [ ] ~~Icon has been created (if new sample) following the [Thumbnail Style Guide and templates](https://github.com/windows-toolkit/WindowsCommunityToolkit-design-assets)~~
- [X] Tests for the changes have been added (for bug fixes / features) (if applicable)
- [X] Header has been added to all new source files (run *build/UpdateHeaders.bat*)
- [X] If new feature, add to [Feature List](https://github.com/windows-toolkit/WindowsCommunityToolkit/blob/master/readme.md#-features)
- [X] Contains **NO** breaking changes

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. 
     Please note that breaking changes are likely to be rejected -->


## Other information
cc. @michael-hawker - would you like me to create a sample/docs page for these two extensions, or is it not needed since they're just two small extensions in an already existing class? Keeping this as a draft PR for now since I'm not sure about this specific point 😄

michael-hawker commented:
@Sergio0694 just going through other issues and saw #1499, can you help me understand how these two things relate?

Sergio0694 commented:
@michael-hawker Those looks like unrelated extensions, if not for the fact that they (indirectly) both work on a target `ScrollViewer` at the end of the day. But for instance, those extensions target the `ListViewBase` type, so they would actually go in another extension class (`ListViewBaseExtensions`?), and not in the one I'm using here. Also, those are just doing manipulations on the internal `ScrollViewer` to bring a given item in view, while the ones in this PR instead are about creating a composition expression animation to get an animation that follows the scrolling on the entire targeted `ScrollViewer`. So yeah, different APIs in use and different use cases. Hope this helps! 😊

michael-hawker commented:
It could be cool to have a sample to show you how you'd setup that sort of syncing as a sample page.

Sergio0694 commented:
Added a sample in the `ScrollViewerExample` sample page, along with a reference C# snippet. Also modified the API surface to add an additional `VisualProperty` option, so that users can choose whether to animate through `Visual.Translation` or `Visual.Offset`.

For instance, `Visual.Translate` is enabled by default as it's the one that generally works fine, but there are cases where `Visual.Offset` might be preferred (eg. when the target items are inside a `Canvas`).

michael-hawker commented on Microsoft.Toolkit.Uwp.SampleApp/SamplePages/ScrollViewerExtensions/ScrollViewerExtensionsXaml.bind:
```diff
@@ -0,0 +1,114 @@
+        </Style>
+      </ListView.ItemContainerStyle>
+    </ListView>
+    <StackPanel x:Name="shapesPanel"
```
I think my main comment is that it wasn't super obvious that this was tied to the code-behind shown in the code panel and what the effect was (as it's pretty seamless outside of the show animation on ListView).

Maybe add a comment here alluding to the helper behind the scenes? Also it could be interesting if you offset the icons (without the backgrounds) ***between*** the ListView items, as then that could show a neat use case that'd be hard to do otherwise without these extensions.

Sergio0694 replied:
Added more docs with a XAML comment, and tweaked the effect in https://github.com/windows-toolkit/WindowsCommunityToolkit/pull/3220/commits/50b455b1822ac7eb8acb19c5b770942b6465b5c4.

michael-hawker commented on Microsoft.Toolkit.Uwp.UI/Extensions/ScrollViewer/ScrollViewerExtensions.cs:
```diff
@@ -11,7 +10,7 @@
 {
     /// <summary>
-    /// Provides attached dependency properties for the <see cref="Windows.UI.Xaml.Controls.ListViewBase"/>
+    /// Provides attached dependency properties for the <see cref="ListViewBase"/>
```
This should be ScrollViewer, no? :P

Sergio0694 replied:
Whoops! 😅

Fixed in https://github.com/windows-toolkit/WindowsCommunityToolkit/pull/3220/commits/31a1b8dd3e554e2686fa1716244389dc618c3a43.

azchohfi replied:
You fixed it only on the ScrollViewerExtensions.ExpressionAnimations.cs class, not on the ScrollViewerExtensions.cs
Would you mind fixing it here as well?

Sergio0694 replied:
Ah, so that's where that typo was originally coming front! 😄

Fixed in https://github.com/windows-toolkit/WindowsCommunityToolkit/pull/3220/commits/5c366a473a514c4679c41999f5471d8849a271ac for all `ScrollViewer` extension classes.

michael-hawker commented:
I'll pull this down and re-review now, but I think we'll be good!

michael-hawker commented:
All good, we should Squash and Merge this one again, as there's the weird history coming from @Sergio0694's branches. Hopefully we can get that reset and fixed in future PRs. 😋
